password_security/models/res_users.py

```diff
@@ -97,22 +98,23 @@ def _check_password(self, password):
         return True
 
     def _check_password_rules(self, password):
-        self.ensure_one()
         if not password:
             return True
-        company_id = self.company_id
-        params = self.env["ir.config_parameter"].sudo()
-        minlength = params.get_param("auth_password_policy.minlength", default=0)
-        password_regex = [
-            "^",
-            "(?=.*?[a-z]){" + str(company_id.password_lower) + ",}",
-            "(?=.*?[A-Z]){" + str(company_id.password_upper) + ",}",
-            "(?=.*?\\d){" + str(company_id.password_numeric) + ",}",
-            r"(?=.*?[\W_]){" + str(company_id.password_special) + ",}",
-            ".{%d,}$" % int(minlength),
-        ]
-        if not re.search("".join(password_regex), password):
-            raise ValidationError(self.password_match_message())
+        users_by_company = groupby(self, lambda u: u.company_id.id)
+        for company_id, users in users_by_company:
+            company_id = users[0].company_id
```

Comment:
This seems unnecessary because it's assigned in the for loop, maybe you just need to group by the record instead of the ID, which would produce a similar result as `_read_group` in Odoo >= v18.